poly/sharing_test.go

```diff
@@ -0,0 +1,544 @@
+ */
+
+// Tests that checks whether a method panics can use this funcition
+func deferTest(t *testing.T, message string) {
+	if r := recover(); r == nil {
+		t.Error(message)
+	}
+}
+
+func producePriPoly(g abstract.Group, k int, s abstract.Secret) *PriPoly {
+	return new(PriPoly).Pick(g, k, s, random.Stream)
+}
+
+func produceTwoPriPoly(g abstract.Group, k int, s abstract.Secret) (*PriPoly,
+	*PriPoly) {
+	return producePriPoly(g, k, s), producePriPoly(g, k, s)
+}
+
+func produceThreePriPoly(g abstract.Group, k int, s abstract.Secret) (*PriPoly,
+	*PriPoly, *PriPoly) {
+	return producePriPoly(g, k, s), producePriPoly(g, k, s),
+		producePriPoly(g, k, s)
+}
+
+func producePriShares(g abstract.Group, k, n int, s abstract.Secret) *PriShares {
+
+	testPoly := producePriPoly(g, k, s)
+	return new(PriShares).Split(testPoly, n)
+}
+
+func producePubPoly(g abstract.Group, k, n int, s abstract.Secret,
+	p abstract.Point) *PubPoly {
+
+	testPriPoly := producePriPoly(g, k, s)
+	testPubPoly := new(PubPoly)
+	testPubPoly.Init(g, n, p)
+	return testPubPoly.Commit(testPriPoly, p)
+}
+
+func producePubShares(g abstract.Group, k, n, t int, s abstract.Secret,
+	p abstract.Point) *PubShares {
+
+	testPubPoly := producePubPoly(g, k, n, s, p)
+	return new(PubShares).Split(testPubPoly, t)
+}
+
+/* Test Functions */
+
+// Test that the Pick function creates unique polynomials and unique secrets.
+func TestPriPolyPick_UniqueShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, nil)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Failed to create unique secrets.")
+	}
+}
+
+// Test polynomials that are based on common secrets. Verify that
+// unique polynomials are created but that the base secrets are all the same.
+func TestPriPolyPick_CommonShares(t *testing.T) {
+	testPoly1, testPoly2, testPoly3 := produceThreePriPoly(group, k, secret)
+	if testPoly1.Equal(testPoly2) || testPoly1.Equal(testPoly3) ||
+		testPoly2.Equal(testPoly3) {
+		t.Error("Failed to create unique polynomials.")
+	}
+	if !testPoly1.Secret().Equal(testPoly2.Secret()) ||
+		!testPoly1.Secret().Equal(testPoly3.Secret()) ||
+		!testPoly2.Secret().Equal(testPoly3.Secret()) {
+		t.Error("Polynomials are expected to have the same secret.")
+	}
+}
+
+// Verify that the secret function works. The function should return the same
+// secret it was initialized with.
+func TestPriPolySecret(t *testing.T) {
+	testPoly := producePriPoly(group, k, secret)
+	if !secret.Equal(testPoly.Secret()) {
+		t.Error("This secret differs from the one given to it.")
+	}
+}
+
+// Verify that Equal returns true for two polynomials that are the same
+func TestPriPolyEqual_Same(t *testing.T) {
```

Comment:
I think in this and a lot of your test functions where you have multiple TestPriPoly<func>_<testcase> functions, it would probably be better and produce more concise and simpler test-code if you just merge all the <testcase> variant functions into a single TestPriPoly<func> function that performs all of the test cases.  In this case, for example, just have a TestPriPolyEqual function that first performs a same-polynomial test and then a different-polynomial test.